### PR TITLE
Add functionality of registration and login

### PR DESCRIPTION
**Pull request by XeniaVi**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,16 +19,20 @@
   },
   "homepage": "https://github.com/XeniaVi/backend-todo#readme",
   "dependencies": {
+    "bcryptjs": "^2.4.3",
     "cors": "^2.8.5",
     "dotenv": "^16.0.1",
     "express": "^4.18.1",
+    "jsonwebtoken": "^8.5.1",
     "mongoose": "^6.4.1",
     "ts-node-dev": "^2.0.0",
```
Comment (nortoza) [inline package.json]: This is dev dependency

File: src/app.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import express from "express";
 import mongoose, { ConnectOptions } from "mongoose";
-import router from "./routers/router";
+import todoRouter from "./routers/todoRouter";
+import authRouter from "./routers/authRouter";
```
Comment (nortoza) [inline src/app.ts]: Lets create index.js file in all folders like routers, controllers, services, helpers, types, etc and import from there

File: src/config/config.ts [hunk 3]
```diff
@@ -5,4 +5,5 @@ dotenv.config();
 export const config = {
   DB_URL: process.env.DB_URL || "localhost",
   PORT: process.env.PORT || 3000,
+  SECRET_KEY: process.env.SECRET_KEY,
```
Comment (nortoza) [inline src/config/config.ts]: Lets add default value

File: src/helpers/helper.ts [hunk 4]
```diff
@@ -9,3 +11,22 @@ export const prepareTodoObject = (item: TodoDB) => {
...
+  const { _id, username, password, roles } = item;
+  return {
+    id: _id.toString(),
+    username,
+    password,
+    roles,
+  };
+};
+
+export const generateAccessToken = (id: string, username: string) => {
```
Comment (nortoza) [inline src/helpers/helper.ts]: Lets unifiy this method - get the payload object from out

File: src/helpers/helper.ts [hunk 5]
```diff
@@ -9,3 +11,22 @@ export const prepareTodoObject = (item: TodoDB) => {
...
+  const payload = {
+    id,
+    username,
+  };
+
+  return jwt.sign(payload, config.SECRET_KEY, { expiresIn: "24h" });
```
Comment (nortoza) [inline src/helpers/helper.ts]: Lets move expires in value to env

File: src/models/User.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import mongoose from "mongoose";
+
+const User = new mongoose.Schema({
+  username: { type: String, unique: true, required: true },
+  password: { type: String, required: true },
+  roles: [{ type: String, ref: "Role" }],
```
Comment (nortoza) [inline src/models/User.ts]: Why roles, not role?

File: src/routers/authRouter.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+const router = Router();
+
+router.post(
+  "/registration",
+  validateRegistrationUsename,
+  validateRegistrationPassword,
+  AuthController.registration
+);
+router.post("/login", AuthController.login);
+router.post("/role", AuthController.makeRole);
```
Comment (nortoza) [inline src/routers/authRouter.ts]: Do we really need this logic?

File: src/validators/authValidators.ts [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+};
+
+export const validateRegistrationPassword = (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  const { password } = req.body;
+
+  if (password.length < 4 || password.length > 15) {
```
Comment (nortoza) [inline src/validators/authValidators.ts]: Lets move this magic values to consts

File: src/validators/authValidators.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
+import { ErrorWrongData } from "../errors/errors";
+import { Request, Response, NextFunction } from "express";
+
+export const validateRegistrationUsename = (
```
Comment (nortoza) [inline src/validators/authValidators.ts]: We can unite these middlewares to one

File: src/services/AuthService.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
+import Role from "../models/Role";
+import User from "../models/User";
+import { ErrorWrongData } from "../errors/errors";
+import bcript from "bcryptjs";
+import { UserDB } from "../types/types";
+import { prepareUserObject, generateAccessToken } from "../helpers/helper";
+
+class AuthService {
+  static async registration(username: string, password: string) {
+    const res = await User.findOne({ username });
```
Comment (nortoza) [inline src/services/AuthService.ts:10]: Lets wrap these methods` logic to try catch

Comment (XeniaVi) [reply src/services/AuthService.ts:10]: I've made, but not at all sure that it is correct

File: src/app.ts [hunk 11]
```diff
@@ -1,17 +1,18 @@
...
 import mongoose, { ConnectOptions } from "mongoose";
-import router from "./routers/router";
+import { todoRouter, authRouter } from "./routers";
 import { config } from "./config/config";
 import cors from "cors";
-import { ErrorWrongData } from "./errors/errors";
-import { typeErrors } from "./constants/constants";
+import { ErrorWrongData } from "./errors";
+import { typeErrors } from "./constants";
 import { Request, Response, NextFunction } from "express";
```
Comment (nortoza) [inline src/app.ts]: Duplicate import

File: src/controllers/AuthController.ts [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    try {
+      const response = await AuthService.getRoles();
+      return res.json(response);
+    } catch (error) {
+      next(error);
+    }
+  }
+}
+
+export default AuthController;
```
Comment (nortoza) [inline src/controllers/AuthController.ts]: Lets export not as default (add above "export class...")

File: src/controllers/index.ts [hunk 13]
```diff
@@ -0,0 +1,4 @@
+import AuthController from "../controllers/AuthController";
```
Comment (nortoza) [inline src/controllers/index.ts]: Then lets export these things as "export * from..."

File: src/middleware/index.ts [hunk 14]
```diff
@@ -0,0 +1,3 @@
+import { AuthMiddleware } from "./AuthMiddleware";
```
Comment (nortoza) [inline src/middleware/index.ts]: The same

File: src/models/index.ts [hunk 15]
```diff
@@ -0,0 +1,5 @@
+import Role from "./Role";
```
Comment (nortoza) [inline src/models/index.ts]: The same

File: src/routers/index.ts [hunk 16]
```diff
@@ -0,0 +1,4 @@
+import todoRouter from "./todoRouter";
```
Comment (nortoza) [inline src/routers/index.ts]: The same

File: src/services/AuthService.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+import { nextTick } from "process";
+
+class AuthService {
+  static async registration(username: string, password: string) {
+    try {
+      const res = await User.findOne({ username });
+
+      if (res) throw new ErrorWrongData("User already exists");
+
+      const hashPassword = bcript.hashSync(password, 7);
```
Comment (nortoza) [inline src/services/AuthService.ts]: Lets move 7 to consts

File: src/services/AuthService.ts [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+      await user.save();
+
+      return { message: "User successfully created" };
+    } catch (error) {
+      return { message: error.message };
+    }
+  }
+
+  static async login(username: string, password: string) {
+    let users: Array<UserDB> = await User.find({ username });
```
Comment (nortoza) [inline src/services/AuthService.ts]: let -> const

File: src/services/AuthService.ts [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+    const validPassword = bcript.compareSync(password, user.password);
+    if (!validPassword) throw new ErrorWrongData(`Password is wrong`);
+
+    const payload = {
+      id: user.id,
+      user: user.username,
+    };
+
+    const token = generateAccessToken(payload);
```
Comment (nortoza) [inline src/services/AuthService.ts]: Just return

File: src/services/index.ts [hunk 20]
```diff
@@ -0,0 +1,4 @@
+import AuthService from "./AuthService";
```
Comment (nortoza) [inline src/services/index.ts]: The same

File: src/validators/authValidators.ts [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+  res: Response,
+  next: NextFunction
+) => {
+  const { username, password } = req.body;
+
+  if (!username) {
+    next(new ErrorWrongData("Username must not be empty"));
+  }
+
+  const min = 4;
```
Comment (nortoza) [inline src/validators/authValidators.ts]: To consts file)

File: src/validators/authValidators.ts [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+  const max = 15;
+
+  if (password.length < min || password.length > max) {
+    next(
+      new ErrorWrongData(
+        "Password's length must be more than 4 and less than 15"
```
Comment (nortoza) [inline src/validators/authValidators.ts]: Here replace it too on variables

File: src/validators/index.ts [hunk 23]
```diff
@@ -0,0 +1,5 @@
+import { validateNewTodo, validateUpdateTodo } from "./todoValidators";
```
Comment (nortoza) [inline src/validators/index.ts]: The same

File: src/middleware/index.ts [hunk 24]
```diff
@@ -0,0 +1 @@
+export * from "./AuthMiddleware";
```
Comment (nortoza) [inline src/middleware/index.ts]: Lets call dir middleware -> middlewares

File: src/services/AuthService.ts [hunk 25]
```diff
@@ -0,0 +1,59 @@
+import { Role } from "../models";
+import { User } from "../models";
```
Comment (nortoza) [inline src/services/AuthService.ts]: Duplicate import

File: src/services/AuthService.ts [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+
+      const user = new User({
+        username,
+        password: hashPassword,
+      });
+      await user.save();
+
+      return { message: "User successfully created" };
+    } catch (error) {
+      return { message: error.message };
```
Comment (igirin) [inline src/services/AuthService.ts:29]: Lets throw this it

File: src/services/AuthService.ts [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+    }
+  }
+
+  static async login(username: string, password: string) {
+    const users: Array<UserDB> = await User.find({ username });
```
Comment (igirin) [inline src/services/AuthService.ts:34]: in this method we can have an error throwed too. Lets wrap it on try catch and other methods too

File: src/services/TodoService.ts [hunk 28]
```diff
@@ -1,9 +1,9 @@
...
-import { ErrorWrongData } from "../errors/errors";
+import { TodoNew, TodoDB, TodoPrepared, TodoUpdate } from "../types";
+import { Todo } from "../models";
+import { prepareTodoObject } from "../helpers";
+import { ErrorWrongData } from "../errors";
 
-class TodoService {
+export class TodoService {
   static async add(todo: TodoNew) {
     const newTodo: TodoDB = await Todo.create(todo);
```
Comment (igirin) [inline src/services/TodoService.ts:8]: try catch

**Pull request merged**